text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/resolver/Formatter.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * This file is part of adventure, licensed under the MIT License.
+ *
+ * Copyright (c) 2017-2022 KyoriPowered
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.minimessage.tag.resolver;
+
+import java.text.DecimalFormat;
+import java.text.DecimalFormatSymbols;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.TemporalAccessor;
+import java.util.Locale;
+import net.kyori.adventure.text.minimessage.tag.Tag;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Tag resolvers producing tags that insert formatted values.
+ *
+ * <p>These are effectively placeholders.</p>
+ *
+ * @since 4.10.0
+ */
+public final class Formatter {
+  private Formatter() {
+  }
+
+  /**
+   * Creates a replacement that inserts a number as a component. The component will be formatted by the provided DecimalFormat.
+   *
+   * <p>This tag expects a format as attribute. Refer to {@link DecimalFormat} for usable patterns.</p>
+   *
+   * <p>You can specify the decimal and grouping separator by adding those as another attribute.</p>
+   *
+   * <p>This replacement is auto-closing, so its style will not influence the style of following components.</p>
+   *
+   * @param key the key
+   * @param number the number
+   * @return the placeholder
+   * @since 4.10.0
+   */
+  public static TagResolver formatNumber(final @NotNull String key, final Number number) {
+    return TagResolver.resolver(key, (argumentQueue, context) -> {
+      final String format = argumentQueue.popOr("Format expected").value();
+      final DecimalFormat decimalFormat = new DecimalFormat(format);
+      final DecimalFormatSymbols formatSymbols = new DecimalFormatSymbols(Locale.ROOT);
```

Comment:
Do we want to be able to provide a locale here? or as an argument to the method? That would allow doing some level of localized number formatting

Reply:
Yea, that would be a good idea. Imo Localisation should be in the tag. E.g. I want to use MiniMessage to parse localized messages for different languages. Passing the Locale to the method wouldn't really help in this case.

Maybe better add it as an argument instead? I'm not sure what's a good syntax for that, I welcome every suggestion.

Reply:
I was sort of thinking of offering some sort of typed map on the parse context that would hold the locale? maybe base off of pointers or something?

That way, you could do sth like:

```java
Component localized(final Player player, final String key) {
  final String message = this.messagesBundle.get(player.locale(), key);
  return this.mm.deserialize(message, player.pointers(), /* any extra meta */);
}
```

or more generally:

```java
Component localized(final Locale lang, final String key) {
  final String message = this.messagesBundle.get(lang, key);
  return this.mm.deserialize(message, Pointers.builder().withStatic(Identity.LOCALE, lang).build(), /* any extra meta */);
}
```

does that seem reasonable? is it too tied in with adventure?

Reply:
I don't think that's a good idea. It might be better to provide the locale in a tag?
Smth like `<lang:DE_de>Number should be formatted with , separator: <num:'#.00'>`?

I'm not sure but maybe there will be multi language text?
E.g. `<lang:EN_en>The translation of the text <blue>The number is <num:'#.00'></blue> in German is:<lang:DE_de><blue> Die nummer ist <num:'#.00></blue>` That might be a edge case but Imo it should be covered by minimessage.